### PR TITLE
Friendships v1

### PR DESCRIPTION
**Pull request by ashraffares**
Add friendships model 
Add friendships method in user
-pending_friends: Get all pending friends
-friend_requests: Get all friends request
-friend?:  Ask if a user is on the friend list

Add friendships controller 
-Add create 
-add destroy
-add accept which accept an invitation to a given user

Alter the show page
-show pending friends
-show friends 
-show friends request
-Add link to approve or decline a friend request

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Changes Required 🙏🏿 
Nice work in this project. You have managed to check most checkboxes. Well done!
Let's make some changes before merging this PR.
- [x] A user should be able to see their friend requests. This is currently not working.
- [x] A user should be able to reject requests.
- [x] A user should be able to send a request to another user from the list of users and from user profiles. Right now one can only see the request link from a user's profile.
- [x] Posts should only be seen by users and their friends. Right now anyone logged in can see a post by other users who they are not friends with. Fix that.
- [x] Add both integration and unit tests


### Martin Afani
#### Slack: whiz25

**Comment by ashraffares** [comment]
@whiz25 
- User can see friends requests
- Only user and friends can see  his post
- Add unfriend and approve or reject friend request in the index
- Add tests

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
**Hello @ashraffares, Well done on working on some changes suggested by @whiz25. Kindly implement the below changes to get this project approved**
- [x] There is a routing error when the decline friendship is clicked.
- [x] when the friendship is confirmed, the button should change to remove friend.
- [x] Kindly ensure that posts can only be seen by people who are friends and not all logged-in users.
- [x] For the unit tests, you have only tested for associations, ensure that you test validations and all public methods as well
- [x] In the user's show page, there is an `unfriend friend` that is generating a routing error when clicked. Kindly fix it
- [x] Kindly add integration tests.

Kindly implement the above changes before submitting another review.
**Feel free to reach me on slack or leave a comment while mentioning me and I will be happy to reply**

_Slack: Oloo Moses_

File: app/views/users/_user.html.erb [hunk 1]
```diff
@@ -1,6 +1,10 @@
 <li>
   Name: <%= user.name %>
   <span class="profile-link">
+  
     <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
+    <%if !user.friend?(current_user) and user != current_user%>
```
Comment (henatan99) [inline app/views/users/_user.html.erb]: This if condition should also consider if the current_user has already sent a friend request for the 'user' which is not yet accepted or declined. The link should show something like 'remove' after a friend request is already sent. The current_user can change his mind and remove the friend request he/she sent. In addition, the current_user can hit the 'Add A Friend' link several times and results in undesirable behavior. You can see the attached screenshot image of your app's behavior. This is also related to friendship association and validation problems in your user model. 

Screenshot1
Screenshot2

**Review by henatan99 — APPROVED** [review]
# Project Approved :100:
Hi there, congratulations , your project is approved :partying_face: :fireworks: :champagne: 
You have done nice work on this project. Please consider the OPTIONAL improvements even though they are not required for submission of your project. 

## To Highlight 
- Nice to implement the scaffolds method 
- RESTfull web app design 
- Good understanding of Rails MVC architecture 

## OPTIONAL improvements 
The comments given here are optional only because they were not mentioned in the previous reviews. I highly recommend you consider them, otherwise, your project would look weak and might impart a wrong impression about your Rails MVC skills.
- The Friendship association has some issues to fix. Currently, your app lets the current_user send a friend request to a user who sent a friend request for the current_user. This is due to an association issue and also needs validation (uniqueness of friendship). See also the inline comment. 
- Your app allows already friends users to send several friend requests. Please test your models and associations in the console using the 'rails c' command and test the db tables. 

### You can now merge the PR and submit the mandatory project submission form. 

If you want to reach out to me, please use the below links:
[LinkedIn](https://www.linkedin.com/in/henok-mekonnen/) | [Twitter](https://twitter.com/henatan99) | Henok Mekonnen on Slack 

## Happy Coding :computer:

**Pull request merged**